test/distributed/fsdp/test_fsdp_summon_full_params.py

```diff
@@ -0,0 +1,278 @@
+
+    @skip_if_lt_x_gpu(2)
+    def test_summon_single_param(self):
+        model = FSDP(nn.Linear(1, 1, bias=False)).cuda(self.rank)
+
+        p = model.get_parameter("_fsdp_wrapped_module.flat_param")
+        self.assertEqual(1, p.numel())
+
+        with torch.no_grad():
+            # This sets the local shard value
+            p[0] = self.rank + 2
+
+        with model._summon_full_params(writeback=True):
+            self.assertEqual(1, p.numel())
+            with torch.no_grad():
+                p.copy_(torch.zeros_like(p))
+
+        # most ranks hold no data and wrote to padding so only rank zero will observe the above write
+        if self.rank == 0:
+            self.assertEqual(0, p[0])
+        else:
+            self.assertEqual(self.rank + 2, p[0])
+
+    @skip_if_lt_x_gpu(2)
+    def test_reshard_outside_forward_backward_iteration(self):
+        model = FSDP(nn.Sequential(
+            FSDP(nn.Linear(5, 5, bias=False)),
+            nn.Linear(5, 1, bias=False)
+        )).cuda(self.rank)
+
+        outer_param = model.get_parameter("_fsdp_wrapped_module.flat_param")
+        inner_param = model.get_parameter("_fsdp_wrapped_module._fpw_module.0._fsdp_wrapped_module.flat_param")
+        outer_full_param_size = outer_param.numel() * self.world_size
+
+        # First lets validate our assumption about resharding
+
+        output = model(torch.zeros(5).cuda(self.rank))
+        # the root FSDP module keeps all params around
+        self.assertEqual(outer_full_param_size, outer_param._full_param_padded.storage().size())
+        self.assertEqual(0, inner_param._full_param_padded.storage().size())
+
+        output.backward()
+        # we reshard everything after backward() finishes
+        self.assertEqual(0, outer_param._full_param_padded.storage().size())
+        self.assertEqual(0, inner_param._full_param_padded.storage().size())
+
+        # now lets repeat it with summon done in between
+
+        output = model(torch.zeros(5).cuda(self.rank))
+        with model._summon_full_params():
+            pass
+        self.assertEqual(outer_full_param_size, outer_param._full_param_padded.storage().size())
+        self.assertEqual(0, inner_param._full_param_padded.storage().size())
+
+        output.backward()
+        with model._summon_full_params():
+            pass
+        self.assertEqual(0, outer_param._full_param_padded.storage().size())
+        self.assertEqual(0, inner_param._full_param_padded.storage().size())
+
+
+    @skip_if_lt_x_gpu(2)
+    def test_params_are_unflatenned(self):
+        model = FSDP(nn.Linear(self.world_size, 1, bias=False)).cuda(self.rank)
+
+        flattened_param = model.get_parameter("_fsdp_wrapped_module.flat_param")
+        self.assertEqual(1, flattened_param.numel())
+
+        with model._summon_full_params():
+            a = model.weight.flatten().detach()
+            b = flattened_param.detach()
+            self.assertTrue(torch.equal(a, b))
```

Comment:
Can be addressed in a follow up diff - I think this tests that if we flatten the params they are equal, but we should ideally also test the original shapes.